### PR TITLE
Do not leak dummy variable in user owned ProjectInstance object

### PR DESCRIPTION
**Pull request by cdmihai**
In addition, put a protective check in the two known msbuild string caches to preemptively avoid caching unique strings.

False positive caching is still hit in the following scenarios
- building successive in-memory mutations of the same `ProjectInstance` instance, in the same BeginBuild-EndBuild-ResetCaches session
- building different `ProjectInstance` instances over the same file, in the same BeginBuild-EndBuild-ResetCaches session

To fix these I think we have to actually remove the dummy variable and change `ConfigurationMetadata` (implements what it means to be a unique build in the engine) to account for the entire `ProjectInstance` state.

### CODE REVIEW

File: src/Build/BackEnd/Shared/BuildRequestConfiguration.cs [hunk 1]
```diff
@@ -490,6 +490,9 @@ internal void CreateUniqueGlobalProperty()
             // create a copy so the mutation does not leak into the ProjectInstance
             _globalProperties = new PropertyDictionary<ProjectPropertyInstance>(_globalProperties);
 
+            // if you change the key also change the following so that it does not leak into string caches:
+            // Microsoft.Build.OpportunisticIntern.BucketedPrioritizedStringList.TryIntern
+            // Microsoft.Build.Construction.ProjectStringCache.Add
             var key = $"ProjectInstance{Guid.NewGuid():N}";
```
Comment (rainersigwald) [inline src/Build/BackEnd/Shared/BuildRequestConfiguration.cs]: Can you define a constant and use it in all 3 places?

File: src/Build/BackEnd/Shared/BuildRequestConfiguration.cs [hunk 2]
```diff
@@ -485,6 +485,18 @@ public void InitializeProject(BuildParameters buildParameters, Func<ProjectInsta
...
 
+        internal void CreateUniqueGlobalProperty()
+        {
+            // create a copy so the mutation does not leak into the ProjectInstance
+            _globalProperties = new PropertyDictionary<ProjectPropertyInstance>(_globalProperties);
+
+            // if you change the key also change the following so that it does not leak into string caches:
+            // Microsoft.Build.OpportunisticIntern.BucketedPrioritizedStringList.TryIntern
+            // Microsoft.Build.Construction.ProjectStringCache.Add
+            var key = $"ProjectInstance{Guid.NewGuid():N}";
```
Comment (rainersigwald) [inline src/Build/BackEnd/Shared/BuildRequestConfiguration.cs]: Should this name be prefixed with `MSBuild` like other reserved/magic properties?

File: src/Shared/OpportunisticIntern.cs [hunk 3]
```diff
@@ -886,6 +886,28 @@ internal void ReportStatistics(string heading)
...
+                             candidate[5] == 'c' &&
+                             candidate[6] == 't' &&
+                             candidate[7] == 'I' &&
+                             candidate[8] == 'n' &&
+                             candidate[9] == 's' &&
+                             candidate[10] == 't' &&
+                             candidate[11] == 'a' &&
+                             candidate[12] == 'n' &&
+                             candidate[13] == 'c' &&
+                             candidate[14] == 'e'
```
Comment (rainersigwald) [inline src/Shared/OpportunisticIntern.cs]: Oh, this probably makes the use-a-constant part hard. 😢

Comment (cdmihai) [inline src/Shared/OpportunisticIntern.cs]: True, but I can iterate over the constant. The compiler hopefully will unroll it a sufficient number of times to actually remove the loop entirely.

Comment (rainersigwald) [inline src/Shared/OpportunisticIntern.cs]: Probably best to check the IL output of the compiler. I suspect Roslyn doesn't do that kind of unrolling, and the [JIT doesn't](https://github.com/dotnet/coreclr/issues/11606).

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for Desktop please

**Pull request merged**